### ISSUE TITLE
Server-sent events (SSE) don't work with CIO server

### ISSUE THREAD

**Issue by jakubkulhan** [ktorio/ktor#1199]
### Ktor Version

1.2.2

### Ktor Engine Used

ktor-server-cio

### JVM Version, Operating System and Relevant Context

```
$ java -version
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)
```

### Feedback

When trying to produce server-sent events, response body doesn't get sent to client. If I start server like this:

```kotlin
package cz.scuk.app

import io.ktor.application.call
import io.ktor.http.CacheControl
import io.ktor.http.ContentType
import io.ktor.response.cacheControl
import io.ktor.response.respondTextWriter
import io.ktor.routing.get
import io.ktor.routing.routing
import io.ktor.server.cio.CIO
import io.ktor.server.engine.embeddedServer
import kotlinx.coroutines.delay

fun main() {
    embeddedServer(CIO, port = 8080) {
        routing {
            get("/") {
                call.response.cacheControl(CacheControl.NoCache(null))
                call.respondTextWriter(contentType = ContentType.Text.EventStream) {
                    while (true) {
                        write("data: ${System.currentTimeMillis()}\n\n")
                        flush()
                        delay(1000)
                    }
                }
            }
        }
    }.start(true)
}
```

And then send request with curl:

```
$ curl -v --no-buffer http://localhost:8080/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: no-cache
< Transfer-Encoding: chunked
< Content-Type: text/event-stream; charset=UTF-8
< 

```

The curl hangs, no response is being received.

When `CIO` implementation is replaced with `Netty`, client periodically receives data chunks. Also when `delay(1000)` is removed, data are also sent to the client.

**Comment by cy6erGn0m** [comment]
It looks like flush() doesn't work in CIO

**Comment by cy6erGn0m** [comment]
This is caused by the chunked encoder that doesn't handle flushes at all and buffers exactly chunk size before sending bytes.

**Comment by cy6erGn0m** [comment]
Could you please describe your use-case: why do you use SSE and not WebSockets?

**Comment by jakubkulhan** [comment]
I need to start local HTTP server in an Android app to create a bridge between a web app and the native app - more specifically, the native app uses SDK for built-in device barcode scanner, when a barcode is scanned, the web app needs to be sent message with scanned data.

I tried to use Ktor WebSockets, however, they depend on `java.time` APIs that aren't available till Android API level 26, so I tried using server-sent events and encountered this issue.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-605">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by arkanovicz** [comment]
> Could you please describe your use-case: why do you use SSE and not WebSockets?

Just out of curiosity, why not?

**Comment by e5l** [comment]
Fixed in 1.5.2
